localstack/services/awslambda/invocation/version_manager.py

```diff
@@ -0,0 +1,386 @@
+    logs: Optional[str] = None
+
+
+@dataclasses.dataclass(frozen=True)
+class LogItem:
+    log_group: str
+    log_stream: str
+    logs: str
+
+
+class ShutdownPill:
+    pass
+
+
+QUEUE_SHUTDOWN = ShutdownPill()
+
+
+class LogHandler:
+    log_queue: "Queue[Union[LogItem, ShutdownPill]]"
+    _thread: Optional[Thread]
+    _shutdown_event: threading.Event
+
+    def __init__(self) -> None:
+        self.log_queue = Queue()
+        self._shutdown_event = threading.Event()
+        self._thread = None
+
+    def run_log_loop(self) -> None:
+        while not self._shutdown_event.is_set():
+            log_item = self.log_queue.get()
+            if log_item is QUEUE_SHUTDOWN:
+                return
+            store_cloudwatch_logs(log_item.log_group, log_item.log_stream, log_item.logs)
+
+    def start_subscriber(self) -> None:
+        self._thread = Thread(target=self.run_log_loop)
+        self._thread.start()
+
+    def add_logs(self, log_item: LogItem) -> None:
+        self.log_queue.put(log_item)
+
+    def stop(self) -> None:
+        self._shutdown_event.set()
+        if self._thread:
+            self.log_queue.put(QUEUE_SHUTDOWN)
+            self._thread.join(timeout=2)
+            if self._thread.is_alive():
+                LOG.error("Could not stop log subscriber in time")
+            self._thread = None
+
+
+class LambdaVersionManager(ServiceEndpoint):
+    # arn this Lambda Version manager manages
+    function_arn: str
+    function_version: FunctionVersion
+    # mapping from invocation id to invocation storage
+    running_invocations: Dict[str, RunningInvocation]
+    # stack of available (ready to get invoked) environments
+    available_environments: "queue.LifoQueue[Union[RuntimeEnvironment, ShutdownPill]]"
+    # mapping environment id -> environment
+    all_environments: Dict[str, RuntimeEnvironment]
+    # queue of invocations to be executed
+    queued_invocations: "Queue[Union[QueuedInvocation, ShutdownPill]]"
+    provisioned_concurrent_executions: int
+    invocation_thread: Optional[Thread]
+    shutdown_event: threading.Event
+    state: VersionState
+    log_handler: LogHandler
+
+    def __init__(
+        self,
+        function_arn: str,
+        function_version: FunctionVersion,
+    ):
+        self.function_arn = function_arn
+        self.function_version = function_version
+        self.running_invocations = {}
+        self.available_environments = queue.LifoQueue()
+        self.all_environments = {}
+        self.queued_invocations = Queue()
+        self.provisioned_concurrent_executions = 0
+        self.invocation_thread = None
+        self.shutdown_event = threading.Event()
+        self.state = VersionState(
+            state=State.Pending, code=StateReasonCode.Creating, reason="Function starting up"
+        )
+        self.log_handler = LogHandler()
+
+    def start(self) -> None:
+        try:
+            invocation_thread = Thread(target=self.invocation_loop)
```

Comment:
nit: we could set `invocation_thread.daemon = True`, to ensure the thread is terminated on process teardown (probably already covered by the logic around `self.shutdown_event`, but generally a good pattern to be on the safe side)